### PR TITLE
Separate quaternion algebra from particle rotation

### PR DESCRIPTION
**Pull request by jngrad**
Description of changes:
- extract quaternion algebra from the `particle_data.hpp` and `rotation.hpp` files
   - removes code duplication caused by a (now resolved) circular dependency (see #3157)
   - makes it possible to replace the quaternion code by a dedicated library, e.g. [boost:qvm::quat](https://www.boost.org/doc/libs/1_68_0/libs/qvm/doc/index.html) or [boost::math::quaternion](https://www.boost.org/doc/libs/1_62_0/libs/math/doc/html/quaternions.html) in the core (see #2289), [rowan](https://rowan.readthedocs.io/en/latest/) in the interface (see #2964)
- simplify code around call sites using Vector4d arithmetic, `std::tuple`, Particle references
- documentation cleanup

### CODE REVIEW

**Comment by RudolfWeeber** [comment]
I'd suggest deferring this until after 4.1. The freeze is already a few days overdue and I'd like to focus on release-relevant issues, right now.

File: src/core/quaternion.hpp [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+  auto const dm = d.norm();
+
+  // null vectors cannot be converted to quaternions
+  if (dm < ROUND_ERROR_PREC) {
+    return {1, 0, 0, 0};
+  }
+
+  // Calculate angles
+  auto const d_xy = sqrt(d[0] * d[0] + d[1] * d[1]);
+  double theta2, phi2;
```
Comment (fweik) [inline src/core/quaternion.hpp]: `T`

File: src/core/quaternion.hpp [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+ *  if the director is the null vector.
+ */
+template <class T>
+Utils::Vector<T, 4>
+convert_director_to_quaternion(Utils::Vector<T, 3> const &d) {
+
+  auto const dm = d.norm();
+
+  // null vectors cannot be converted to quaternions
+  if (dm < ROUND_ERROR_PREC) {
```
Comment (fweik) [inline src/core/quaternion.hpp]: This assumes `dm` is `double`, please use `std::numeric_limits`.

File: src/core/quaternion.hpp [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    // Here the director is co-linear with the z-azis
+    // We need to distinguish between (0, 0, +d_z) and (0, 0, -d_z)
+    theta2 = (d[2] > 0) ? 0 : Utils::pi() / 2;
+    phi2 = 0;
+  } else {
+    // Here we take care of all other directions
+    // We suppose that theta2 = theta/2 and phi2 = (phi - pi/2)/2,
+    // where angles theta and phi are in spherical coordinates
+    theta2 = acos(d[2] / dm) / 2;
+    phi2 = ((d[1] > 0) ? 1 : -1) * acos(d[0] / d_xy) / 2 - Utils::pi() / 4;
```
Comment (fweik) [inline src/core/quaternion.hpp]: `Utils::sgn(d[1])`, `Utils::pi<T>()`

File: src/core/quaternion.hpp [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    return {1, 0, 0, 0};
+  }
+
+  // Calculate angles
+  auto const d_xy = sqrt(d[0] * d[0] + d[1] * d[1]);
+  double theta2, phi2;
+  if (d_xy == 0) {
+    // Here the director is co-linear with the z-azis
+    // We need to distinguish between (0, 0, +d_z) and (0, 0, -d_z)
+    theta2 = (d[2] > 0) ? 0 : Utils::pi() / 2;
```
Comment (fweik) [inline src/core/quaternion.hpp]: `Utils::pi<T>()`

File: src/core/quaternion.hpp [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+  You should have received a copy of the GNU General Public License
+  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+#ifndef QUATERNION_H
+#define QUATERNION_H
+/** \file
+ *  Quaternion algebra.
+ */
+
+#ifdef ROTATION
```
Comment (fweik) [inline src/core/quaternion.hpp]: This is not needed, please remove. (Also observe that the header is not self-contained, because the config is not included).

**Review by fweik — CHANGES_REQUESTED** [review]
Please move `quaternion.hpp` to utils and add unit tests for the three functions in there.

File: src/core/quaternion.hpp (hunk 1 above)
Comment (jngrad) [reply src/core/quaternion.hpp]: thanks!

File: src/core/quaternion.hpp (hunk 2 above)
Comment (jngrad) [reply src/core/quaternion.hpp]: fixed

File: src/core/quaternion.hpp (hunk 3 above)
Comment (jngrad) [reply src/core/quaternion.hpp]: `Utils::sgn(d[1])` won't work here because it returns a value in the set {-1, 0, 1}. If `d[1]` is zero `phi2` will be -pi/2 instead of +pi/2 and the resulting quaternion will have wrong signs, causing the python test `collision_detection` to fail.

File: src/core/quaternion.hpp (hunk 5 above)
Comment (jngrad) [reply src/core/quaternion.hpp]: fixed

**Comment by jngrad** [comment]
@fweik I personally wouldn't move that file to utils because it's not re-usable. The corner case with +/-pi/2 shows this quaternion implementation only works because the calling sites were tailored for it. Writing a C++ unit test would not really make sense here, because only the final result of the rotation is of interest (already covered by Python tests), and it would make maintenance harder (we once had a C++ tabulated bond with the wrong sign, and a test that made sure the result had the wrong sign, because the python logic was tailored for it). Coming up with a unit test is also annoying because the quaternion formalism is not standard, and the fact that a [unit test was hard-coded inside a calling site](https://github.com/espressomd/espresso/blob/61026c799a93e10e2e0900b43b5abb785c69dbdb/src/core/virtual_sites.cpp#L119-L125) to check the rotation inversion implementation is correct every time the function is called is odd.

**Comment by fweik** [comment]
Please add the tests.

**Comment by fweik** [comment]
Testing the multiplication is easy, just check the multiplication table. For the conversion functions you should check that they are inverse to each other. It's sufficient that they encode the direction, not how. If anything, when there is client code that depends on a specific behavior of the function, especially it is non-standard, that gives more reason to add a test, so that behavior is kept. So you should also add a test for the sign convention.
Also the file should be moved, it's generic math and not related to MD or the espressos core, so it is a better fit for utils.

File: src/utils/tests/quaternion_test.cpp [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+  /* self-consistency */
+  using Utils::convert_quaternion_to_director;
+  for (int i = -2; i < 3; ++i) {
+    for (int j = -2; j < 3; ++j) {
+      for (int k = -2; k < 3; ++k) {
+        if (i == 0 and j == 0 and k == 0) {
+          continue;
+        }
+        // here the case where j is zero is also tested
+        Vector3d input{{(double)i, (double)j, (double)k}};
```
Comment (KaiSzuttor) [inline src/utils/tests/quaternion_test.cpp]: ```static_cast<double>(i)```...

Comment (jngrad) [reply src/utils/tests/quaternion_test.cpp]: thanks!

File: src/core/quaternion.hpp (hunk 3 above)
Comment (jngrad) [reply src/core/quaternion.hpp]: The signum edge case is now documented and tested.

**Review by fweik — APPROVED** [review]

**Comment by jngrad** [comment]
bors r=fweik
